R/biting_process.R

```diff
@@ -18,30 +18,78 @@ create_biting_process <- function(
   parameters
   ) {
   function(timestep) {
-    # Calculate combined EIR
     age <- get_age(variables$birth$get_values(), timestep)
-
-    bitten_humans <- simulate_bites(
-      renderer,
-      solvers,
-      models,
+    prob_infection <- get_prob_infection(
       variables,
-      events,
-      age,
+      simulate_bites(
+        renderer,
+        solvers,
+        models,
+        variables,
+        events,
+        age,
+        parameters,
+        timestep
+      ),
       parameters,
       timestep
     )
-
+    prob_asymptomatic_progression <- get_prob_progression(
+      'A',
+      variables$state,
+      1 / parameters$da,
+      parameters
+    )
+    prob_subpatent_progression <- get_prob_progression(
+      'U',
+      variables$state,
+      1 / parameters$du,
+      parameters
+    )
+    prob_outcome <- cbind(
+      prob_infection,
+      prob_asymptomatic_progression,
+      prob_subpatent_progression
+    )
+    outcomes <- simulate_competing_outcomes(prob_outcome)
     simulate_infection(
       variables,
       events,
-      bitten_humans,
+      individual::Bitset$new(
+        parameters$human_population
+      )$insert(which(outcomes == 1)),
       age,
       parameters,
       timestep,
       renderer
     )
+    events$subpatent_progression$schedule(which(outcomes == 2), 0)
+    events$recovery$schedule(which(outcomes == 3), 0)
+  }
+}
+
+#' @importFrom stats runif 
+simulate_competing_outcomes <- function(prob_outcome) {
+  prob_occurs <- apply(prob_outcome, 1, function(p) {
+    1 - prod(1 - p)
+  })
+  occurs <- runif(nrow(prob_outcome)) < prob_occurs
+  outcomes <- rep(NA, nrow(prob_outcome))
+  n_occurances <- sum(occurs)
+  if (n_occurances == 0) {
+    return(outcomes)
+  }
+
+  if (n_occurances == 1) {
+    # apply doesn't work for vectors in the same way
+    p <- prob_outcome[occurs, ]
+    outcomes[occurs] <- sample.int(n = length(p), size = 1, prob = p)
+  } else {
```

Comment:
You can force this to keep the output as a matrix, even if `n_occurances = 1`, by adding `,drop = FALSE`
Example:
```
x <- matrix(1:4, ncol = 2)
# Turns to vector
x[c(TRUE, FALSE),]
# Keep as matrix
x[c(TRUE, FALSE), ,drop = FALSE]
```
then I think you wouldn't need this additional if statement